### PR TITLE
kv_hash to return result for range checks instead of panicking and generalize hasher used

### PR DESCRIPTION
**Pull request by MavenRain**
As the title states, this PR addresses two TODOs related to the kv_hash function.  In the first, it changes the behaviour from panicking on a failed range check to returning an error.  It also allows for different hashes to be used with the same function.  This is achieved by changing the function signature from

```
fn kv_hash(key: &[u8], value: &[u8]) -> Hash
```
to
```
fn kv_hash<D: Digest>(key: &[u8], value: &[u8]) -> Result<Hash, TryFromIntError>
```

### CODE REVIEW

File: src/merk/restore.rs
Comment (fmorency) [inline src/merk/restore.rs:111]: Should this TODO be fixed before merging?

File: src/proofs/tree.rs
Comment (fmorency) [inline src/proofs/tree.rs:41]: nit: Can we simplify this code using https://doc.rust-lang.org/std/result/#comparison-operators ?

**Review by fmorency — APPROVED** [review]
Thanks! Approved, assuming comments are addressed.

File: src/proofs/tree.rs
Comment (MavenRain) [reply src/proofs/tree.rs:41]: @fmorency Quite possibly . . . good catch . . . will have a look

File: src/merk/restore.rs
Comment (MavenRain) [reply src/merk/restore.rs:111]: @fmorency No.  I want to do a little more investigation into this particular TODO to see if it is as simple as it sounds.

File: src/proofs/tree.rs
Comment (MavenRain) [reply src/proofs/tree.rs:41]: @fmorency Unfortunately, we can't leverage `PartialEq` for `Result` because `ed::Error` does not implement `PartialEq`.

**Review by hansl — APPROVED** [review]

**Pull request merged**
